aptos-move/aptos-vm/src/transaction_arg_validation.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+//! Module defines validation of transaction arguments.
+//!
+//! TODO: we should not only validate the types but also the actual values, e.g.
+//! for strings whether they consist of correct characters.
+
+use crate::{
+    move_vm_ext::{MoveResolverExt, SessionExt},
+    VMStatus,
+};
+use move_deps::{
+    move_core_types::{account_address::AccountAddress, value::MoveValue, vm_status::StatusCode},
+    move_vm_runtime::session::LoadedFunctionInstantiation,
+    move_vm_types::loaded_data::runtime_types::Type,
+};
+use once_cell::sync::Lazy;
+use std::collections::BTreeSet;
+
+static ALLOWED_STRUCTS: Lazy<BTreeSet<String>> = Lazy::new(|| {
+    ["0x1::ASCII::String"]
+        .iter()
+        .map(|s| s.to_string())
+        .collect()
+});
+
+/// Validate and generate args for entry function
+/// validation includes:
+/// 1. return signature is empty
+/// 2. number of signers is same as the number of senders
+/// 3. check arg types are allowed after signers
+///
+/// after validation, add senders and non-signer arguments to generate the final args
+pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    senders: Vec<AccountAddress>,
+    args: Vec<Vec<u8>>,
+    func: &LoadedFunctionInstantiation,
+) -> Result<Vec<Vec<u8>>, VMStatus> {
+    // entry function should not return
+    if !func.return_.is_empty() {
+        return Err(VMStatus::Error(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE));
+    }
+    let mut signer_param_cnt = 0;
+    // find all signer params at the beginning
+    for ty in func.parameters.iter() {
+        match ty {
+            Type::Signer => signer_param_cnt += 1,
+            Type::Reference(inner_type) => {
+                if matches!(&**inner_type, Type::Signer) {
```

Comment:
Rookie Rust question: Why the complex &**?

Reply:
I guess since this code c&p from the aptos_vm: The inner type is `&Arc<_>`. So you make an Arc of that. Then the strange `&*` in Rust is an idiom to access the content of a box as a reference (`*` first gets the content out). There is probably a nicer way via `inner_type.as_ref()` but as said, not code I've written.